content/posts/replacing-the-redisgraph-parser.md

```diff
@@ -0,0 +1,109 @@
+---
+title: "Lessons learned while replacing the RedisGraph parser"
+date: 2019-11-22
+authors:
+  - Jeffrey Lovitz
+tags: ["redisgraph", "c", "cypher", "parsers"]
+---
+
+![parser](/parser.png)
+
+Since its inception, RedisGraph used [Flex](https://github.com/westes/flex) and [Lemon](https://www.sqlite.org/lemon.html) (a yacc-like parser generator developed by the SQLite team) to build its query parser. We maintained our own grammar, extending it to accommodate new features as we wrote them. Flex and Lemon are marvelous and valuable tools, but looking back, my take on our initial do-everything-ourselves approach is best captured by Douglas Adams:
+
+> In the beginning the Universe was created.This has made a lot of people very angry and been widely regarded as a bad move.
+
+[59 commits, ~20,000 modified lines](https://github.com/RedisGraph/RedisGraph/pull/488), and a few months later, we've paid back a solid amount of technical debt and learned a few things along the way! RedisGraph queries are fully converted to ASTs by [libcypher-parser](https://github.com/cleishm/libcypher-parser), an Apache-licensed project authored and maintained over the past 4 years by Chris Leishman (on [Twitter](https://twitter.com/cleishm) and [Github](https://github.com/cleishm/) as cleishm).
+
+## When to repay technical debt
+
+> The extra effort that it takes to add new features is the interest paid on the debt.
+
+To use [Martin Fowler's language of technical debt](https://martinfowler.com/bliki/TechnicalDebt.html), our self-maintained grammar charged us interest every time we introduced or modified a syntactical construct. Gradual improvement was not an option, however, so we were reluctant to undertake the effort of a full replacement. We finally relented when working on the WITH clause, which is how Cypher allows for multiple queries to be chained into a single atomic execution. Unlike subqueries in SQL, which create a tree of independent executions, the WITH clause flattens query parts into a linear composition:
+
+![Multi-Part Query Railroad Diagram](/multipart_query.png)
+[Cypher railroad diagram for multi-part queries](https://s3.amazonaws.com/artifacts.opencypher.org/railroad/MultiPartQuery.html)
+
+Our original grammar was effective at interpreting a sequence of clauses as a self-contained query, but it would have required a major rewrite to interpret that same sequence as one scope capable of reading from and projecting into others. libcypher-parser had already solved this problem and many more, so we started the even more ambitious rewrite of scrapping our parser and building off an open-source project.
+
+## The joy of FOSS
+libcypher-parser was a perfect fit for our needs. Beyond the core criteria of being a feature-rich Cypher parser, it is a C implementation, stable but evolving, and - above all - open-source.
```

Comment:
change to: "Beyond the core criteria of being a feature-rich Cypher parser, it is a C implementation, stable but evolving, and —above all – open source."